File: pkgs/ffigen/test/native_objc_test/runtime_version_test.dart
// Copyright (c) 2022, the Dart project authors. Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

// Objective C support is only available on mac.
@TestOn('mac-os')
import 'dart:ffi';
import 'dart:io';

import 'package:objective_c/objective_c.dart';
import 'package:path/path.dart' as path;
import 'package:test/test.dart';

import '../test_utils.dart';
import 'runtime_version_bindings.dart';
import 'util.dart';

void main() {
  group('runtime version check', () {
    setUpAll(() {
      // TODO(https://github.com/dart-lang/native/issues/1068): Remove this.
      DynamicLibrary.open(
        path.join(
          packagePathForTests,
          '..',
          'objective_c',
          'test',
          'objective_c.dylib',
        ),
      );
      final dylib = File(
        path.join(
          packagePathForTests,
          'test',
          'native_objc_test',
          'objc_test.dylib',
        ),
      );
      verifySetupFile(dylib);
      DynamicLibrary.open(dylib.absolute.path);
      generateBindingsForCoverage('runtime_version');
    });

    test('Interface', () {
      expect(() => FutureAPIInterface(), throwsA(isA<OsVersionError>()));
    });

    test('Interface methods', () {
      final api = FutureAPIMethods();
      expect(() => api.unavailableMac(), throwsA(isA<OsVersionError>()));
      expect(api.unavailableIos(), 2);
      expect(() => api.unavailableBoth(), throwsA(isA<OsVersionError>()));
      expect(() => api.futureMethodMac(), throwsA(isA<OsVersionError>()));
      expect(api.futureMethodIos(), 5);
      expect(() => api.futureMethodBoth(), throwsA(isA<OsVersionError>()));
    });

    test('Category methods', () {
      final api = NSObject();
      expect(() => api.catUnavailableMac(), throwsA(isA<OsVersionError>()));
      expect(api.catUnavailableIos(), 2);
      expect(() => api.catUnavailableBoth(), throwsA(isA<OsVersionError>()));
      expect(() => api.catFutureMethodMac(), throwsA(isA<OsVersionError>()));
      expect(api.catFutureMethodIos(), 5);
      expect(() => api.catFutureMethodBoth(), throwsA(isA<OsVersionError>()));
    });
  });
}

File: pkgs/ffigen/test/native_objc_test/runtime_version_test.m
#import <Foundation/NSObject.h>

API_DEPRECATED("test", ios(1000.0, 1001.0), macos(1000.0, 1001.0))
@interface FutureAPIInterface : NSObject;
@end

@implementation FutureAPIInterface : NSObject;
@end

@interface FutureAPIMethods : NSObject;
-(int)unavailableMac API_UNAVAILABLE(macos);
-(int)unavailableIos API_UNAVAILABLE(ios);
-(int)unavailableBoth API_UNAVAILABLE(ios, macos);
-(int)futureMethodMac API_DEPRECATED("test", macos(1000.0, 1001.0));
-(int)futureMethodIos API_DEPRECATED("test", ios(1000.0, 1001.0));
-(int)futureMethodBoth API_DEPRECATED("test", macos(1000.0, 1001.0), ios(1000.0, 1001.0));
@end

@implementation FutureAPIMethods;
-(int)unavailableMac { return 1; }
-(int)unavailableIos { return 2; }
-(int)unavailableBoth { return 3; }
-(int)futureMethodMac { return 4; }
-(int)futureMethodIos { return 5; }
-(int)futureMethodBoth { return 6; }
@end

@interface NSObject (FutureAPICategoryMethods)
-(int)catUnavailableMac API_UNAVAILABLE(macos);
-(int)catUnavailableIos API_UNAVAILABLE(ios);
-(int)catUnavailableBoth API_UNAVAILABLE(ios, macos);
-(int)catFutureMethodMac API_DEPRECATED("test", macos(1000.0, 1001.0));
-(int)catFutureMethodIos API_DEPRECATED("test", ios(1000.0, 1001.0));
-(int)catFutureMethodBoth API_DEPRECATED("test", macos(1000.0, 1001.0), ios(1000.0, 1001.0));
@end

@implementation NSObject (FutureAPICategoryMethods)
-(int)catUnavailableMac { return 1; }
-(int)catUnavailableIos { return 2; }
-(int)catUnavailableBoth { return 3; }
-(int)catFutureMethodMac { return 4; }
-(int)catFutureMethodIos { return 5; }
-(int)catFutureMethodBoth { return 6; }
@end
